sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala

```diff
@@ -111,14 +111,25 @@ object AgnosticEncoders {
 
   object ProductEncoder {
     val cachedCls = new ConcurrentHashMap[Int, Class[_]]
-    def tuple(encoders: Seq[AgnosticEncoder[_]]): AgnosticEncoder[_] = {
+    private[sql] def tuple(
```

Comment:
TBH I'd just create a bespoke method for this for joinWith. The current signature is pretty confusing IMO.